google-cloud-bigquery/test/google/cloud/bigquery/dataset/routines_test.rb

```diff
@@ -0,0 +1,183 @@
+
+    first_routines.count.must_equal 3
+    first_routines.each { |ds| ds.must_be_kind_of Google::Cloud::Bigquery::Routine }
+    first_routines.next?.must_equal true
+
+    second_routines.count.must_equal 2
+    second_routines.each { |ds| ds.must_be_kind_of Google::Cloud::Bigquery::Routine }
+    second_routines.next?.must_equal false
+  end
+
+  it "paginates routines with all" do
+    mock = Minitest::Mock.new
+    mock.expect :list_routines, list_routines_gapi(dataset.dataset_id, 3, "next_page_token"),
+      [project, dataset.dataset_id, max_results: nil, page_token: nil]
+    mock.expect :list_routines, list_routines_gapi(dataset.dataset_id, 2, nil),
+      [project, dataset.dataset_id, max_results: nil, page_token: "next_page_token"]
+    dataset.service.mocked_service = mock
+
+    routines = dataset.routines.all.to_a
+
+    mock.verify
+
+    routines.count.must_equal 5
+    routines.each { |ds| ds.must_be_kind_of Google::Cloud::Bigquery::Routine }
+  end
+
+  it "paginates routines with all and max" do
+    mock = Minitest::Mock.new
+    mock.expect :list_routines, list_routines_gapi(dataset.dataset_id, 3, "next_page_token"),
+      [project, dataset.dataset_id, max_results: 3, page_token: nil]
+    mock.expect :list_routines, list_routines_gapi(dataset.dataset_id, 2, nil),
+      [project, dataset.dataset_id, max_results: 3, page_token: "next_page_token"]
+    dataset.service.mocked_service = mock
+
+    routines = dataset.routines(max: 3).all.to_a
+
+    mock.verify
+
+    routines.count.must_equal 5
+    routines.each { |ds| ds.must_be_kind_of Google::Cloud::Bigquery::Routine }
+  end
+
+  it "iterates routines with all using Enumerator" do
+    mock = Minitest::Mock.new
+    mock.expect :list_routines, list_routines_gapi(dataset.dataset_id, 3, "next_page_token"),
+      [project, dataset.dataset_id, max_results: nil, page_token: nil]
+    mock.expect :list_routines, list_routines_gapi(dataset.dataset_id, 3, "second_page_token"),
+      [project, dataset.dataset_id, max_results: nil, page_token: "next_page_token"]
+    dataset.service.mocked_service = mock
+
+    routines = dataset.routines.all.take(5)
+
+    mock.verify
+
+    routines.count.must_equal 5
+    routines.each { |ds| ds.must_be_kind_of Google::Cloud::Bigquery::Routine }
+  end
+
+  it "iterates routines with all with request_limit set" do
+    mock = Minitest::Mock.new
+    mock.expect :list_routines, list_routines_gapi(dataset.dataset_id, 3, "next_page_token"),
+      [project, dataset.dataset_id, max_results: nil, page_token: nil]
+    mock.expect :list_routines, list_routines_gapi(dataset.dataset_id, 3, "second_page_token"),
+      [project, dataset.dataset_id, max_results: nil, page_token: "next_page_token"]
+    dataset.service.mocked_service = mock
+
+    routines = dataset.routines.all(request_limit: 1).to_a
+
+    mock.verify
+
+    routines.count.must_equal 6
+    routines.each { |ds| ds.must_be_kind_of Google::Cloud::Bigquery::Routine }
+  end
```

Comment:
possibly want an list test in this test set that leverages filter once it is available.